perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -4109,7 +4109,14 @@ public Group createGroupUnion(PerunSession sess, Group resultGroup, Group operan
 		// check if there is already a record of these two groups
 		if (this.groupsManagerImpl.isRelationBetweenGroups(resultGroup, operandGroup)) {
 			throw new GroupRelationAlreadyExists("There is already a relation defined between result group " + resultGroup +
-					" and operand group " + operandGroup + " or they are in hierarchical structure.");
+					" and operand group " + operandGroup + " or they are in direct hierarchical structure.");
+		}
+
+		// check indirect relationships (for example "A" with "A:B:C" are in indirect relationship using "A:B")
+		List<Group> allParentGroups = this.getParentGroups(sess, operandGroup);
+		if(allParentGroups.contains(operandGroup)) {
```

Comment:
Shouldn't this verify the presence of the resultGroup instead?
```suggestion
		if(allParentGroups.contains(resultGroup)) {
```

Reply:
You are right, this would be always true. I checked wrong group. Fixed.